### PR TITLE
Uc002 login

### LINKED ISSUE #15: 🔐 UC002 - Login
### User Story

**As a** salesman with an account on the app
**I want** to log in with my email and password
**In order to** access the application

### Acceptation Test

**Given** a salesman
**When** he arrives on the app 
**Then** the option to log in the application is proposed to him

**Given** a salesman on the login page who gave his correct credentials
**When** he clicks on the "login" button
**Then** he is logged in and can access the app

**Given** a salesman on the login page who gave incorrect credentials
**When** he clicks on the "login" button
**Then** his connection request is refused and he is informed of the denial reason

**Given** a salesman on the login page who gave correct credentials but doesn't have an internet connection
**When** he clicks on the "login" button
**Then** he is informed that the creation isn't possible due to a connection error

### PR DESCRIPTION
**Pull request by Francois389**
close #15 

# Result

image

If both field are empty, a popup appear
image

If the email isn't well formed, a popup appear
image

### CODE REVIEW

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/MainActivity.java [hunk 1]
```diff
@@ -1,24 +1,77 @@
...
 import android.os.Bundle;
+import android.view.View;
+import android.widget.EditText;
+import android.widget.TextView;
 
 import androidx.activity.EdgeToEdge;
 import androidx.appcompat.app.AppCompatActivity;
 import androidx.core.graphics.Insets;
 import androidx.core.view.ViewCompat;
 import androidx.core.view.WindowInsetsCompat;
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/MainActivity.java:13]: Missing Javadoc comment to describe what the activity is for

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/Login.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+import androidx.appcompat.app.AppCompatActivity;
+
+import fr.iut_rodez.pathpilot_android_client.MainActivity;
+import fr.iut_rodez.pathpilot_android_client.R;
+import fr.iut_rodez.pathpilot_android_client.util.Network;
+import fr.iut_rodez.pathpilot_android_client.util.Popup;
+import fr.iut_rodez.pathpilot_android_client.util.ValidateForm;
+
+public class Login extends AppCompatActivity {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/Login.java:18]: Missing Javadoc comment to describe what the activity is for

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.Date;
+
+import fr.iut_rodez.pathpilot_android_client.BuildConfig;
+import fr.iut_rodez.pathpilot_android_client.util.Network;
+import fr.iut_rodez.pathpilot_android_client.util.Popup;
+
+public class LoginService {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:31]: Missing Javadoc comment to describe what the activity is for

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+                    progressDialog.dismiss();
+
+                    handleError(context, error);
+                }) {
+        };
+
+        requestQueue.add(request);
+    }
+
+    private static void saveAuthToken(String token, Context context) {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:79]: Don't forget signature (no need for getter/setter/constructor, only for class)

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        // Utilisez SharedPreferences avec chiffrement ou un gestionnaire de tokens sécurisé
+        SharedPreferences sharedPreferences = context.getSharedPreferences("AuthPrefs", Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = sharedPreferences.edit();
+        editor.putString("auth_token", token);
+        editor.apply();
+        Log.d(TAG, "saveAuthToken: Token saved");
+        Log.d(TAG, "saveAuthToken: Token: " + token);
+    }
+
+    public static class LoginInput {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:89]: Missing Javadoc comment to describe what the class is for

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+                json.put("password", password);
+            } catch (JSONException e) {
+                // This should never happen, has the keys are Strings
+                Log.e(TAG, "Error while converting LoginInput to JSON", e);
+            }
+            return json;
+        }
+    }
+
+    public static class LoginResponse {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:111]: Missing Javadoc comment to describe what the class is for

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Network.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import android.net.ConnectivityManager;
+import android.net.NetworkInfo;
+
+import com.android.volley.RequestQueue;
+import com.android.volley.toolbox.Volley;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Network {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Network.java:14]: Missing Javadoc comment to describe what the class is for

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Popup.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
+package fr.iut_rodez.pathpilot_android_client.util;
+
+import android.app.AlertDialog;
+import android.content.Context;
+import android.widget.Toast;
+
+public class Popup {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Popup.java:7]: Missing Javadoc comment to describe what the class is for

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+import com.android.volley.ParseError;
+import com.android.volley.ServerError;
+import com.android.volley.TimeoutError;
+import com.android.volley.VolleyError;
+
+import java.nio.charset.StandardCharsets;
+
+import fr.iut_rodez.pathpilot_android_client.R;
+
+public class VolleyErrorHandler {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java:21]: Missing Javadoc comment to describe what the class is for

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+    public static String handleError(Context context, VolleyError error) {
+        String errorMessage = "Unknown error occurred";
+
+        // Check network connectivity first
+        if (!Network.isNetworkConnected(context)) {
+            return "No internet connection. Please check your network settings.";
+        }
+
+        // Detailed error handling
+        if (error instanceof NetworkError) {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java:40]: Use of the enhanced switch more readable

```java
String errorMessage = switch (error) {
    case NetworkError ne -> "Network error. Unable to connect to the server.";
    case ServerError se -> handleServerError(se);
    case AuthFailureError ae -> "Authentication failed. Please give valid credentials.";
    case ParseError pe -> "Unable to process server response.";
    case TimeoutError te -> "Connection timed out. Please try again.";
    default -> "An unexpected error occurred.";
};```

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+        return errorMessage;
+    }
+
+    /**
+     * Handles server-specific error messages
+     *
+     * @param error Volley error
+     * @return Specific server error message
+     */
+    private static String handleServerError(VolleyError error) {
```
Comment (Odonata971) [inline app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java:68]: Same here with enhanced switch like this 

```java
return switch (day) {
            case MONDAY, TUESDAY, WEDNESDAY, THURSDAY, FRIDAY -> "Weekday";
            case SATURDAY, SUNDAY -> "Weekend";
        };
```

File: app/src/main/res/drawable/leave_icon.xml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
+    android:width="125dp"
+    android:height="102dp"
+    android:viewportWidth="125"
+    android:viewportHeight="102">
+    <path
+        android:pathData="M122.711 56.1455C125.763 53.2996 125.763 48.6777 122.711 45.8317L91.4595 16.6888C88.4076 13.8429 83.4513 13.8429 80.3994 16.6888C77.3475 19.5348 77.3475 24.1567 80.3994 27.0027L98.3202 43.7143H46.8773C42.5558 43.7143 39.0644 46.9701 39.0644 51C39.0644 55.0299 42.5558 58.2857 46.8773 58.2857H98.3202L80.3994 74.9973C77.3475 77.8433 77.3475 82.4652 80.3994 85.3112C83.4513 88.1572 88.4076 88.1572 91.4595 85.3112L122.711 56.1683V56.1455ZM39.0644 14.5714C43.3859 14.5714 46.8773 11.3156 46.8773 7.28571C46.8773 3.2558 43.3859 0 39.0644 0H23.4386C10.4986 0 0 9.79018 0 21.8571V80.1429C0 92.2098 10.4986 102 23.4386 102H39.0644C43.3859 102 46.8773 98.7442 46.8773 94.7143C46.8773 90.6844 43.3859 87.4286 39.0644 87.4286H23.4386C19.1171 87.4286 15.6258 84.1728 15.6258 80.1429V21.8571C15.6258 17.8272 19.1171 14.5714 23.4386 14.5714H39.0644Z"
```
Comment (Odonata971) [inline app/src/main/res/drawable/leave_icon.xml:7]: i think there is an error on 7:84

File: app/src/test/java/fr/iut_rodez/pathpilot_android_client/util/ValidateFormTest.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
+package fr.iut_rodez.pathpilot_android_client.util;
+
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Test;
+
+public class ValidateFormTest {
```
Comment (Odonata971) [inline app/src/test/java/fr/iut_rodez/pathpilot_android_client/util/ValidateFormTest.java:9]: Missing Javadoc comment to describe what the activity is for

**Review by Odonata971 — CHANGES_REQUESTED** [review]

File: app/src/main/res/drawable/leave_icon.xml (hunk 12 above)
Comment (Francois389) [reply app/src/main/res/drawable/leave_icon.xml:7]: Fix in ` `

File: app/src/test/java/fr/iut_rodez/pathpilot_android_client/util/ValidateFormTest.java (hunk 13 above)
Comment (Francois389) [reply app/src/test/java/fr/iut_rodez/pathpilot_android_client/util/ValidateFormTest.java:13]: Fix in `3f381088a74825447a65980cc8bb98e00dd60954`

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java (hunk 9 above)
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java:20]: Fix in `3f381088a74825447a65980cc8bb98e00dd60954`

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:92]: Fix in [3f38108](https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/3f381088a74825447a65980cc8bb98e00dd60954)

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java:20]: Fix in [3f38108](https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/3f381088a74825447a65980cc8bb98e00dd60954)

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java (hunk 11 above)
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java:64]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/aabf7d0cb168a6b98d443e23b957c88983e1d813

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:80]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/3f381088a74825447a65980cc8bb98e00dd60954

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:26]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/3f381088a74825447a65980cc8bb98e00dd60954

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/Login.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/Login.java:22]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/3f381088a74825447a65980cc8bb98e00dd60954

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java (hunk 4 above)
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:80]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/3f381088a74825447a65980cc8bb98e00dd60954

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java (hunk 3 above)
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:26]: FIx in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/3f381088a74825447a65980cc8bb98e00dd60954

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/Login.java (hunk 2 above)
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/Login.java:22]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/3f381088a74825447a65980cc8bb98e00dd60954

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Popup.java (hunk 8 above)
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Popup.java:11]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/eb823270691dc0c912b801466992d2ce62dfca4e

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Network.java (hunk 7 above)
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Network.java:17]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/eb823270691dc0c912b801466992d2ce62dfca4e

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java (hunk 6 above)
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:117]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/eb823270691dc0c912b801466992d2ce62dfca4e

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Popup.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/Popup.java:11]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/eb823270691dc0c912b801466992d2ce62dfca4e

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/login/LoginService.java:117]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/eb823270691dc0c912b801466992d2ce62dfca4e

File: app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+            errorMessage = "Network error. Unable to connect to the server.";
+        } else if (error instanceof ServerError) {
+            errorMessage = handleServerError(error);
+        } else if (error instanceof AuthFailureError) {
+            errorMessage = "Authentication failed. Please give valid credentials.";
+        } else if (error instanceof ParseError) {
+            errorMessage = "Unable to process server response.";
+        } else if (error instanceof TimeoutError) {
+            errorMessage = "Connection timed out. Please try again.";
+        }
```
[the remark this answers is not in the document]
Comment (Francois389) [reply app/src/main/java/fr/iut_rodez/pathpilot_android_client/util/VolleyErrorHandler.java]: Fix in https://github.com/SAE-S5-01-Developpement-avance/PathPilot-android-client/pull/38/commits/aabf7d0cb168a6b98d443e23b957c88983e1d813

**Review by Odonata971 — APPROVED** [review]
LGTM

**Pull request merged**